### PR TITLE
Additional MatchRanking Test

### PR DESCRIPTION
**Pull request by hadleyshapland**
I added a new test that checks randomly generated bios up to 5,000 words to improve the test coverage.

### CODE REVIEW

File: src/test/java/com/google/sps/MatchRankingTest.java [hunk 1]
```diff
@@ -331,4 +366,33 @@ private void testOnlyAddNewBio(String newBio, User user) {
     user.setBioMap(savedMatchWordCount);
   }
 
+  private String getRandomGoodBio(int bioLength) {
+      String[] possibleWords = {"security", "engineer", "software", "Security", "data", "work", "secure", "engineering", "work", "google", "microsoft", "structures",
+      "working", "network", "securing", "endpoints", "cloud", "android", "app", "application"};
+
+      Random random = new Random();
+      String toReturn = "";
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/MatchRankingTest.java:374]: Consider using a StringBuilder which should be more efficient when appending strings in a loop.

File: src/test/java/com/google/sps/MatchRankingTest.java [hunk 2]
```diff
@@ -331,4 +366,33 @@ private void testOnlyAddNewBio(String newBio, User user) {
     user.setBioMap(savedMatchWordCount);
   }
 
+  private String getRandomGoodBio(int bioLength) {
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/MatchRankingTest.java:369]: Would it be useful to create a more generic function that can generate a bio using a mixture of words from different collections? This could allow you to create fake bios that have higher or lower keyword frequencies. 

I was thinking something like the following:
```
   private String generateRandomBio(int targetLength, float keywordPercentage, Collection<String> keywords, Collection<String> nonKeywords);
```

File: src/test/java/com/google/sps/MatchRankingTest.java (hunk 1 above)
Comment (hadleyshapland) [reply src/test/java/com/google/sps/MatchRankingTest.java]: Done

File: src/test/java/com/google/sps/MatchRankingTest.java (hunk 2 above)
Comment (hadleyshapland) [reply src/test/java/com/google/sps/MatchRankingTest.java]: This was a really good idea, I updated the function to take in keyword percentage so that I had more flexibility with the tests

File: src/test/java/com/google/sps/MatchRankingTest.java [hunk 3]
```diff
@@ -313,8 +318,68 @@ private static boolean compareDoubles(double d1, double d2) {
...
+        List<User> result = MatchRanking.rankMatches(currentUser.getBioMap(), newMatches);
+        Assert.assertEquals(userGood, result.get(0));
+      } catch (Exception e) {
+        System.err.println("System could not handle bio of length " + i + ". See testDifferentLengthBios for more details.");
+        break;
+      }
+    }
+  }
+
+  //prints the percent keywords that pass with bio ranked at the top
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/MatchRankingTest.java:351]: Nice work! 

For this test, the goal would be to validate that the rankMatches method does not get worse with future code changes. Could we also add some assertions based on the percentage?

File: src/test/java/com/google/sps/MatchRankingTest.java [hunk 4]
```diff
@@ -313,8 +318,68 @@ private static boolean compareDoubles(double d1, double d2) {
...
+      userGood.setBio(generateRandomBio(50, i));
+      userBad.setBio(generateRandomBio(50, 0.5));
+
+      Collection<User> newMatches = new HashSet<User>();
+      newMatches.add(userGood);
+      newMatches.add(userBad);
+
+      List<User> result = MatchRanking.rankMatches(currentUser.getBioMap(), newMatches);
+
+      if(result.get(0).equals(userGood)) {
```
Comment (ghandre) [inline src/test/java/com/google/sps/MatchRankingTest.java:375]: Somewhat mirroring what Jerry put, these tests are close to what we'd call benchmark tests.

Building, running and maintaining benchmarks usually is done with custom tooling that stores various performance metrics from previous runs (memory use, cpu usage, thread contention, time taken, etc) and reevaluates it on various machines and platforms. This is of course out of scope for the work here.

What I'd suggest instead is focusing on the match performance (as you are, by checking the two users rank correctly), but maybe try and plot out (e.g.: with google sheets) what the relationship is between `userGood` being ranked higher than `userBad` and visa-versa with respect to the percentage `i`.

For instance, I'd imagine the keyword-percentage-i relates to the ranking result sort of like the following table:
i | userGood before userBad
0 | 0 <- In 0% of runs, userGood appeared before userBad when it had no keywords
...
0.45 | 0.30 <- in 30% of runs (just a guess),  userGood appeared before userBad when it had 45% keywords
0.50 | 0.50 <- in 50% of runs userGood appeared before userBad when it had 50% keywords
0.55 | 0.70 <- in 70% of runs (just a guess),  userGood appeared before userBad when it had 55% keywords
...
1 | 1 <- ...

Once you've characterized that relationship, you should be able to use these figures to set up assertions on the expected performance of the match ranking subject to different criterias (values of i). Note that in measuring this for the above table, but also for building the assertions, you probably want to run each keyword-percentage-i multiple times so you get a fraction of success, rather than a binary success/fail.

For the assertion, you could either hard-code the minimum expected performance associated with each i (with some tolerance -- E.g.: if with i=0.50 you'd expect 'goodUser' to outrank 'badUser' 50% of the time, you should set the threshold in the assertion to allow a bit lower than 50%, since random variation will have it move up and down from that on different runs).

TL;DR: This test is good and already adds value! To complete its purpose as a guard against performance regressions / as a benchmark:
- Try running each percentage value multiple (e.g.: 20-100 times, as performance permits), and measure how well it performs as a fraction of successfull and failed rankings
- Hard code those fractions in an assertion (run for each i), so if the code is later broken it issues a test failure

Happy to discuss this further if the above isn't clear :)

One last note: Of course for these purposes synthetic data is all we have; Ordinarily we'd run this on real and recent data, using some expert/better system to establish the ground truth.

File: src/test/java/com/google/sps/MatchRankingTest.java (hunk 3 above)
Comment (hadleyshapland) [reply src/test/java/com/google/sps/MatchRankingTest.java:351]: Done

File: src/test/java/com/google/sps/MatchRankingTest.java (hunk 4 above)
Comment (hadleyshapland) [reply src/test/java/com/google/sps/MatchRankingTest.java]: I updated the test to do this, and also added an output file that shows the exact percentages (I picked lower percentages for the benchmarks, but thought it would be useful to print the actual percents to a file as well). Let me know what you think!

File: src/test/java/com/google/sps/MatchRankingTest.java [hunk 5]
```diff
@@ -361,18 +366,53 @@ public void testDifferentPercentKeywordBios() {
...
 
-    //test different percentages of keywords against a 0% keyword bio
-    for(double i = 0.05; i <= 1.0; i += 0.05) {
-      userGood.setBio(generateRandomBio(50, i));
-      userBad.setBio(generateRandomBio(50, 0.0));
+    //put minimum benchmark checks in map (maps keyword percent to percent it should be ranked first)
+    Map<Integer, Double> expectedPerformance = new HashMap<Integer, Double>();
+    expectedPerformance.put(0, 0.0);
+    expectedPerformance.put(25, 0.0);
+    expectedPerformance.put(50, 0.2);
```
Comment (ghandre) [inline src/test/java/com/google/sps/MatchRankingTest.java:373]: These values are fine (no action requested); Out of curiosity, were these based off the output - some constant, or did you use some theory/empirical evaluation to pick values that would consistently pass the test?

If you're interested / for future reference, a more concrete way of figuring out how much lower than the expected value you should use is to use the properties of the binomial:

Imagine you run one of your cases 1000 times (in order to calibrate), and find that it passes (has the right ranking order) 70% of the time. Then, suppose that, in the test (not just for calibration) you perform 100 repetitions; On average you'd expect to see 70 of them pass, however if you look at the binomial distribution, you'll see that there's still a ~2% probability that you might get as few as 60 passing (here's a [binomial calculator](https://www.wolframalpha.com/widgets/view.jsp?id=78baf4f3a070cc5b9b226664d2ce80ec), I got the 2% by plugging in n=100,  p=0.7, X>60, and then subtracting 100% - 0.97011 (calculator result)) . 

You can then define an acceptable failure rate and use that to define the lower bound (e.g.: perhaps you're ok with the test failing 1:100000 times / a 99.99% success rate -- "four nines" as we'd call it), which in the above example works out to 51 (that is, in 1:100000 cases the ranking order will be correct less than 51 times, causing the test to fail if we use this as the threshold).

That's one way of dealing with pseudo-random tests. In general though, this kind of statistical reasoning is hidden away by the benchmarking tool, and in regular tests we mostly try to avoid making tests random exactly because properly testing semi-random behavior is hard (though sometimes inevitable / better than the alternatives, like here where we don't have real data to work with).

File: src/test/java/com/google/sps/MatchRankingTest.java [hunk 6]
```diff
@@ -361,18 +366,53 @@ public void testDifferentPercentKeywordBios() {
...
+    expectedPerformance.put(100, 0.99);
+
+    FileWriter output = new FileWriter("src/test/java/com/google/sps/matchRankingBenchmark.txt");
+
+    //test different percentages of keywords against a 50% keyword bio
+    for(double i = 0.0; i <= 1.0; i += 0.05) {
+        int rankedFirstCount = 0;
+
+        //run test at each percent level 50 times to get accurate benchmark
+        for(int j = 0; j < 50; ++j) {
```
[the remark this answers is not in the document]
Comment (hadleyshapland) [reply src/test/java/com/google/sps/MatchRankingTest.java:387]: Done, good idea!

File: src/test/java/com/google/sps/matchRankingBenchmark.txt [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+When user had 65% keywords, it ranked ahead of user with half keywords 80% of runs.
+
+When user had 70% keywords, it ranked ahead of user with half keywords 88% of runs.
+
+When user had 75% keywords, it ranked ahead of user with half keywords 98% of runs.
+
+When user had 80% keywords, it ranked ahead of user with half keywords 98% of runs.
+
+When user had 85% keywords, it ranked ahead of user with half keywords 92% of runs.
```
[the remark this answers is not in the document]
Comment (hadleyshapland) [reply src/test/java/com/google/sps/matchRankingBenchmark.txt:35]: That's true, I was trying to make it more realistic to have some words not match :)

File: src/test/java/com/google/sps/MatchRankingTest.java (hunk 5 above)
Comment (hadleyshapland) [reply src/test/java/com/google/sps/MatchRankingTest.java:373]: I got the values by running the test a few times then picking percents a little bit under what I got for the output to make sure it would almost always pass. That's an interesting way of calculating it - I did find that if I ran the test a bunch of times it would occasionally get a really low percentage and fail, so I tried to make the benchmarks low enough that this wouldn't happen too often :)
